engine/src/main/java/org/terasology/recording/EventSystemReplayImpl.java

```diff
@@ -0,0 +1,728 @@
+
+        @Override
+        public void invoke(EntityRef entity, Event event) {
+            try {
+                Object[] params = new Object[2 + componentParams.size()];
+                params[0] = event;
+                params[1] = entity;
+                for (int i = 0; i < componentParams.size(); ++i) {
+                    params[i + 2] = entity.getComponent(componentParams.get(i));
+                }
+                if (!activity.isEmpty()) {
+                    PerformanceMonitor.startActivity(activity);
+                }
+                try {
+                    methodAccess.invoke(handler, methodIndex, params);
+                } finally {
+                    if (!activity.isEmpty()) {
+                        PerformanceMonitor.endActivity();
+                    }
+                }
+            } catch (Exception ex) {
+                logger.error("Failed to invoke event", ex);
+            }
+        }
+
+        @Override
+        public int getPriority() {
+            return priority;
+        }
+
+        @Override
+        public ComponentSystem getHandler() {
+            return handler;
+        }
+    }
+
+    private static class ReceiverEventHandlerInfo<T extends Event> implements EventSystemReplayImpl.EventHandlerInfo {
+        private EventReceiver<T> receiver;
+        private Class<? extends Component>[] components;
+        private int priority;
+
+        ReceiverEventHandlerInfo(EventReceiver<T> receiver, int priority, Class<? extends Component>... components) {
+            this.receiver = receiver;
+            this.priority = priority;
+            this.components = Arrays.copyOf(components, components.length);
+        }
+
+        @Override
+        public boolean isValidFor(EntityRef entity) {
+            for (Class<? extends Component> component : components) {
+                if (!entity.hasComponent(component)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        @Override
+        public void invoke(EntityRef entity, Event event) {
+            receiver.onEvent((T) event, entity);
+        }
+
+        @Override
+        public int getPriority() {
+            return priority;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (obj == this) {
+                return true;
+            }
+            if (obj instanceof EventSystemReplayImpl.ReceiverEventHandlerInfo) {
+                EventSystemReplayImpl.ReceiverEventHandlerInfo other = (EventSystemReplayImpl.ReceiverEventHandlerInfo) obj;
+                if (Objects.equal(receiver, other.receiver)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hashCode(receiver);
+        }
+
+        @Override
+        public Object getHandler() {
+            return receiver;
+        }
+    }
+
+
+}
```

Comment:
I think this will be the main api for event recording. Can you set up a test class and write tests for the core logic of this system? Try to write tests for features, e.g. what happens when the system processes? Must not be perfect in the first step but I'd like to see at least some tests for a new concept ;)

Reply:
Echoing the test potential here, if that hasn't already happened - I do see some other test classes, not sure if they were there when this was originally posted :-)

There are several internal classes in this file and it is quite sizable already - would it make sense to split them out so this class can be shrunk down a bit? Maybe that would also make testing etc easier? Could just dump it all into a sub-package together if needed.